### PR TITLE
Fixes #24294 - add snapshot tests for templates

### PR DESCRIPTION
**Pull request by kamils-iRonin**
This is about snapshot tests for templates. This PR includes:

- rake tasks to generate snapshots
- test that compare rendered templates to the snapshots

### CODE REVIEW

**Comment by timogoebel** [comment]
ok to test

**Comment by timogoebel** [comment]
@lzap, @ares: Mind taking a look? These are the snapshot tests @lzap invented and we talked about.

**Comment by ares** [comment]
Great stuff, I like it very much. I won't be able to review next week. Long term, I'd like to drop list of templates in seeds, and simply go over app/views/unatttended/**/*.erb as we already have all data in metadata. Could you think of other way to select, what templates should be tested? Also is my understanding correct, that if plugin adds such key, the plugin template could also be tested? If so, what is the way to add snapshot to verify against?

Could we also add ptable and report template in order to catch regressions in rendering of other template kinds?

File: lib/tasks/snapshots.rake
Comment (lzap) [inline lib/tasks/snapshots.rake]: I'd probably go for `snap.txt` extension, I can imagine that *.snap files will get opened in weird stuff on Windows or MacOS.

File: lib/tasks/snapshots.rake
Comment (lzap) [inline lib/tasks/snapshots.rake]: I just started reading, but this rake task is missing key task named "check" or "verify" or whatever. It would render all templates and compare that with snapshots, showing diff and failing out in case there is a difference found. Then we can run this from Jenkins easily.

File: test/unit/foreman/renderer/snapshots/PXELinux/AutoYaST_default_PXELinux.snap
Comment (lzap) [inline test/unit/foreman/renderer/snapshots/PXELinux/AutoYaST_default_PXELinux.snap]: This snap is missing end newline, there are more, I am gonna create a PR.

**Comment by lzap** [comment]
I am good, it's a good start indeed. Go ahead and create the testing task and add it to `lib/tasks/jenkins.rake` so we can start using that on the CI. Make sure the test task will also catch edge cases - template is missing or unexpected (has no snapshot).

@ares I think that changing from list to automatic walking of directories should actually happen *after* this is merged, because this feature can actually find if we are doing it wrong (e.g. template with wrong name, template missing or extra).

File: test/unit/foreman/renderer/template_snapshot_service.rb
Comment (lzap) [inline test/unit/foreman/renderer/template_snapshot_service.rb]: I'd prefer CentOS 7.0 here if we have a fixture for that, latest stable version of Red Hat is the one we should be watching for. Ideally all OSes but that's just too slow to test.

File: lib/tasks/snapshots.rake
Comment (timogoebel) [reply lib/tasks/snapshots.rake]: The tests are run as part of the normal unit test suite, we don't need a dedicated rake task for that.

**Comment by ares** [comment]
> @ares I think that changing from list to automatic walking of directories should actually happen after this is merged, because this feature can actually find if we are doing it wrong (e.g. template with wrong name, template missing or extra).

Sure, I just put a question if there's a way that in future, we could accommodate building snaps automatically. But now when I think about it, the SeedHelper could just do Dir.glob and return the same list so it should work out of the box.

I'd still like to see more snaps than just for provisioning templates and I'd like to understand, how to use this from plugins. None of them blocks this PR of course.

**Comment by timogoebel** [comment]
> I'd still like to see more snaps than just for provisioning templates and I'd like to understand, how to use this from plugins. None of them blocks this PR of course.

Yeah, we should add both. The way I see it, plugins need the ability to add templates to the seeding process. I know Katello has it's own code, so does foreman_rescue and foreman_userdata. Not sure about other plugins.

Yesterday, I played with actually having a dedicated test for every template (so you see more easily which one failed). This was tricky to accomplish because I just wanted to run the seeding step once (because it's pretty slow). Unfortunately, we automatically add all fixtures to all tests and this overrides the seeded data. I came up with [a workaround](https://github.com/theforeman/foreman_userdata/blob/9ffaa1cf13fec64263e8a15d71e4416492e30cf4/test/unit/foreman_userdata/renderer_test.rb). Please let me know, what you think and if we should make these changes here as well. @ares, @lzap, @kamils-iRonin, @lucasm-iRonin.

File: test/unit/foreman/renderer/template_snapshot_service.rb
Comment (tbrisker) [inline test/unit/foreman/renderer/template_snapshot_service.rb]: Wouldn't using factory bot here be simpler than manually setting all these attributes?

File: test/unit/foreman/renderer/template_snapshot_service.rb
Comment (tbrisker) [inline test/unit/foreman/renderer/template_snapshot_service.rb]: `Template.where(name: template_names)` would work just as well i think

**Comment by lzap** [comment]
@timogoebel why do you need to have all the templates seeded? It's slow. Doesn't the new renderer allow reading templates from other sources than database? So my idea is to implement a simple "loader" from ERB files directly and rendering that.

**Comment by timogoebel** [comment]
@lzap: We just did not think of this. But I like the idea. @kamils-iRonin: What do you think? We could just implement this similar to the tar archive variant and use this for the snapshot tests. I‘m also thining if we should just move the tar archive plugin to core as well. But that would naturally be a separate PR.

**Comment by kamils-iRonin** [comment]
@lzap @timogoebel That's a good point, loading templates from ERB files directly would be faster. We can create a `SnapshotsConfig` class which will contain a list of files, and make possible to add files to the list by plugins. Something similar to the `allowed_helpers` configuration.

**Comment by timogoebel** [comment]
Actually, we know the files and can generate the filename from the template name reliably. Why do we need the registry? To know what templates should be snapshot tested? If we add a plugin API, we should use that so that plugins can easily seed templates as well.

**Comment by lzap** [comment]
Absolutely, getting rid of the registry is something we want and @ares mentioned above. But we can postpone registry refactoring into another PR if you prefer. This PR just need a fast loading of templates, seeding is *super slow* and we can't merge this (I assume the test will take a minute or more).

**Comment by kamils-iRonin** [comment]
@timogoebel @lzap To know what templates should be snapshot tested we can add a `snapshot_test` key to the template metadata. And then go through all the files inside `app/views/unattended/` and if there is `snapshot_test` set to true, test this template. WDYT?

**Comment by lzap** [comment]
Of course metadata is ideal. I'd default to do the snapshot test so it would be opt-out only for those we can't test.

**Comment by ares** [comment]
I don't think dev info belongs to metadata

**Comment by kamils-iRonin** [comment]
> I don't think dev info belongs to metadata

We can add a new "dev info" section

**Comment by timogoebel** [comment]
> I don't think dev info belongs to metadata
> 
> We can add a new "dev info" section

Why don't we just add a yamlfile with paths to the templates that we want to have tested. This is not super automatic (which I generally prefer), but it's easy to implement and get's this to a point where we can actually merge this. The tests aren't helpful if they only exist in this PR.

**Comment by kamils-iRonin** [comment]
@timogoebel yamlfile is fine with me as well

**Comment by ares** [comment]
Separate yaml works for me. I think we should avoid metadata because they are stored in production DB in case templates are imported.

File: lib/tasks/snapshots.rake
Comment (timogoebel) [inline lib/tasks/snapshots.rake]: Maybe `scope = Foreman::Renderer.get_scope(host: host, source: source)`?

**Review by timogoebel — COMMENTED** [review]
@kamils-iRonin: Nice, I have one small comment. Otherwise: Looks good to me.

File: test/unit/foreman/renderer/template_snapshot_service.rb
Comment (kamils-iRonin) [reply test/unit/foreman/renderer/template_snapshot_service.rb]: there is only CentOS 5.4 in the fixtures

File: test/unit/foreman/renderer/template_files.yaml
Comment (lzap) [inline test/unit/foreman/renderer/template_files.yaml]: A comment "Edit this file and then run rake snapshot..." would be useful here, just a nitpick if you intent to rebase.

File: test/unit/foreman/renderer/template_snapshot_service.rb
Comment (lzap) [reply test/unit/foreman/renderer/template_snapshot_service.rb]: We are in process of getting rid of fixtures, this is 3 years process or even more, we slowly refactor to FactoryBot. Is there any problem with simply not using fixtures for snapshoting host stub? I think it should be easy simply to use FactoryBot for everything (except perhaps Settings that's acceptable).

**Review by lzap — CHANGES_REQUESTED** [review]
The rake task does not work for me, it generates nothing. I deleted some snapshot files randomly and it does not write them again.

Also, I'd like to avoid as many fixtures as we could. Simply construct the fake host via FactoryBot. I think you still need fixtures for settings but that's pretty much if.

**Comment by kamils-iRonin** [comment]
@lzap I now load only settings fixtures and use FactoryBot to build a fake host. To generate snapshots run `bundle exec rake snapshots:generate RAILS_ENV=test`

File: test/unit/foreman/renderer_test.rb
Comment (lzap) [inline test/unit/foreman/renderer_test.rb:12]: Can you create a big comment here something like:

```ruby
#
# This test tries to render all templates mentioned in template_files.yaml
# and compares the result with copies in test/unit/foreman/renderer/snapshots.
# After review of changes, snapshots can be easily regenerated with:
#
#   bundle exec rake snapshots:generate RAILS_ENV=test
#
```

I am pretty sure people will be hitting this and not knowing what is happening. I wish this could be printed to the bottom of unit test suite but unsure how to do this, this would be enough as people tend to open test files first to see what is going on first :-)

File: test/unit/foreman/renderer/template_files.yaml
Comment (lzap) [inline test/unit/foreman/renderer/template_files.yaml]: Nitpicks: No tail endline. But more importantly, I was thinking if `snapshots.yaml` is better name. Look how it lists on most systems:

```
# find test/unit/foreman/renderer/
test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/UserData default.snap.txt
test/unit/foreman/renderer/snapshots/ProvisioningTemplate/user_data/Kickstart default user data.snap.txt
test/unit/foreman/renderer/template_files.yaml
```

To me it should be more obvious that the YAML file belongs to the "snapshots" directory as it provides list for them.

**Review by lzap — CHANGES_REQUESTED** [review]
Well done! I have few nipticks, but the most important issue is that @ares synced templates from community repo yesterday and it currently fails - you need to do new snapshots actually.

I would like to ask if you could rebase on top of this one as well: https://github.com/theforeman/foreman/pull/6052 we are just about to merge it today and we will be cherry picking it.

But all and all - great feature thank you so much.

**Comment by kamils-iRonin** [comment]
@lzap I've applied suggestions and rebased with develop :-)

**Review by lzap — APPROVED** [review]
Thank you!

**Comment by lzap** [comment]
[test foreman]

**Comment by mmoll** [comment]
[test foreman]
[test update]

**Comment by lzap** [comment]
HostJS tests fails, that's expected. Ran the following locally, all good:

* SeedsTest.test_0006_doesn't add a template back that was renamed
* SeedsTest.test_0005_doesn't add a template back that was deleted

My PR is pending for these random errors. Finally, Katello test looks unrelated:

* Katello::Api::Rhsm::CandlepinDynflowProxyControllerIntegrationTest.test_0001_params are not parsed in the controller

Thanks a bunch this is a test I've been hoping for. Merging.

**Pull request merged**
